CIPs/cip-0032.md

```diff
@@ -0,0 +1,117 @@
+
+We will introduce a new `AttestationSlasher` contract which allows slashing of stake if a completed attestation hasn't been observed in the past `B` blocks. To achieve this, we will upgrade the `Attestations` contract to register the selection and completion events in the new `AttestationSlasher` contract. The `AttestationSlasher` will contain the following state to track the latest block number that an attestation was requested/completed on:
+
+```js
+// Tracks the last time an attestation was requested from an issuer (AS node)
+mapping(address => uint256) public lastAttestationRequested;
+// Tracks the last time an attestation was completed by an issuer
+mapping(address => uint256) public lastAttestationCompleted;
+// Tracks the last time an issuer was last slashed
+mapping(address => uint256) public lastSlashed;
+```
+
+A slashing event may occur when the slashing initiator calls `slash()` with the slashable address. Before slashing, the slashing contract checks the following:
+
+1. The validator has been elected continuously for the past `B` blocks.
+2. The issuer was selected for at least one attestation in the past `B` blocks.
+3. The validator wasn't slashed by the `AttestationSlasher` within the past `B` blocks.
+4. The address hasn't completed an attestation in the past `B` blocks.
+5. The addres hasn't submitted `SignalOnline` event in the last `S` blocks. (see [below](#signal-online)).
+
+If the validator was intermittently elected, the contract will only consider the past `B` blocks in which the validator was *continuously* elected. For example, if a validator with a downed attestation node was elected for .5 `B` blocks, was unelected, and re-elected for .5 `B` blocks, then it will not be eligible for slashing since the slasher restarts counting when the validator was re-elected. Additionally, if the validator was elected but was never selected to provide an attestation in the entire period of `B` blocks, then they are not eligible for slashing.
+
+### Signal online
+
+In the unlikely event that a validator was never selected for an attestation during the period of `B` blocks, it's possible for a malicious actor to request multiple attestations with the goal of randomly getting paired with this validator and intentionally leaving the attestation incomplete. To prevent this, we're building into the AS the ability to signal to the `AttestationSlasher` that the service is still online if a slashing event becomes close by signing and submitting a transaction to the slasher contract. This requires that the AS signer key to be minimally funded to pay the gas fees of the transaction, however this adds more piece of mind for validators to be sure they won't be incorrectly slashed. Additionally, we'll be adding two new Prometheus events to the AS to alert a validator when the slashing period is nearly reached and another event to alert when the AS signer key's balance is low.
```

Comment:
What would trigger this signal? Does the attestation service pro-actively send this transaction and how is it monitoring its status of being "slashable"?

Reply:
Also how do we make this resistant to fake "signals", a service that just sends a signed `SignalOnline` regularly?

Reply:
The service would query the new slasher contract which tracks the last successful attestation. If the time period becomes too close to the slashable event, then the attestation service will automatically submit a `SignalOnline` transaction if the account is funded. 

We will need to create a dashboard to monitor the use of this signal. If it looks like a validator is using it as a way to bypass running a healthy service, then we'll need to either:
1. Remove the feature. This feature is a temporary safety mechanism in case we have low onboarding rates. If we have collective confidence in the min-bar, then we can remove this safety feature.
2. Add mitigations such as introducing an increasing financial penalty for subsequent signals or by decreasing the time that signal will last for (lowering `S`) per user.

Reply:
Under what conditions would the attestation service send and non-send this signal? Does attestation service itself maintain a timer to know when its close to a slashable event and check if it's had a completion in that time period?

I guess my hesitancy is if this would negate most slashable events (eg. the two you found in your query above appear to be running attestation services, so they would probably send a signal...)

I think we have a conservative min-bar suggested and we can decrease as we feel confident, so not sure if this is needed. Plus, it adds more complexity. WDYT?

Reply:
Or can we just make sure we are running the attestation bot in mainnet at a rate similar to the slash event rate to ensure that each validator at least gets a chance for 1 non-malicious request

Reply:
Good points Amy. I agree that this feature does add complexity and the benefits could be realized using other mechanisms. Since its purpose is to mitigate the risks until we build up more confidence in the slashing rules, we can get the same benefits by simply lowering the slashable penalty to a trivial amount to begin with. Once everyone is confident we can ramp it up. WDYT?

Reply:
Ya, I like that approach 👍